doc/compiler/osr/OSR.md

```diff
@@ -20,26 +20,232 @@ OpenJDK Assembly Exception [2].
+Therefore, depending on the event, OSR yield points usually include all or a subset of GC points.
+The reason why we would like to limit the set of OSR yield points, if possible, is the cost of the bookkeeping associated with each such point as well as the restrictions that they impose on the optimizer.
 
-# potentialOSRPointHelper
-potentialOSRPointHelper is a call that serves as a marker of places where OSR transition is supported. The target bytecode index to be executed in the interpreter can be obtained by adding an offset (an information stored in the call node, the size of OSR point's bytecode in post-execution OSR) to the bytecode index of the helper call node.
+The actual OSR happens at **OSR Induction points**. It can happen at the time of the yield or later, as described in the following sections.
 
-The helper is not a must for OSR, but it can help locating the OSR points that are available at ILGen but disappear due to later transformations. It works as a bookkeeping tool, to track places that are OSR possible.
+Therefore, at a high level, the sequence of events can be described as follows:
 
-# osrFearPointHelper
-This helper marks a place that has been optimized with runtime assumptions and requires protection of OSR mechnism. This means an OSR point is required between the invalidation point and the fear point.
+       VM event => OSR yield => OSR induction => OSR transition
 
-Similar to potentialOSRPointHelper, this helper is not a must for optimizations requiring OSR mechnism. It works as a tool to help locating the actual fear (being part of the program that may be removed or optimized to something else).
+
+# Voluntary and Involuntary OSR
+
+The difference between voluntary vs. involuntary OSR is whether it is induced by the JIT or the VM. In the case of involuntary OSR, the transition is induced by the VM by forcibly transferring control from the VM code to an OSR handler in the compiled code. Such possible transfer of control is modelled as exception control flow in the IL because it behaves similarly. Therefore, every OSR yield point is also an OSR induction point.
+
+Under voluntary OSR, the JIT controls when to trigger an OSR transition by inserting OSR inducing code into the IL.
+
+Each compiled method is compiled in either of the two modes: all possible OSR transitions from that method are voluntary or all OSR transitions are involuntary.
+OMR provides infrastructure for involuntary OSR. Downstream projects can implement voluntary OSR.
+
+# Inducible and Uniducible OSR Yield Points
+
+During voluntary OSR, the JIT can induce OSR after practically every OSR yield point and those points are called **inducible OSR yield points**. However, there are some OSR yield points after which the JIT cannot induce OSR. Those are referred to as **uninducible OSR yield points**. A typical example of an uninducible OSR yield point is when there is a thunk archetype present on the inline stack.
```

Comment:
>Do you think it would be useful to have that sort of placeholder document with reminders of things that need documenting, or is there already such a place?

I'm not aware of such placeholder document, however, https://github.com/eclipse-openj9/openj9/blob/master/doc/compiler/README.md is the Table Of Contents for all compiler documents, which also includes all documents in OMR. It'd be nice to track things there. Some options I can think of are

1. Create an entry for `thunk archetype`, or entries for other topics in that [TOC](https://github.com/eclipse-openj9/openj9/blob/master/doc/compiler/README.md) and add a TODO next to it. 
2. Create an entry for `thunk archetype`, or entries for other topics in that [TOC](https://github.com/eclipse-openj9/openj9/blob/master/doc/compiler/README.md) and link the entry to a placeholder document `doc/compiler/NeedsDocumentation.md`
3. Create an entry for `thunk archetype`, or entries for other topics in that [TOC](https://github.com/eclipse-openj9/openj9/blob/master/doc/compiler/README.md) and link the entry to issues that are created to track the documentation work. 

I think (1) is probably easier and less overhead. 

The [TOC](https://github.com/eclipse-openj9/openj9/blob/master/doc/compiler/README.md) also contains a link to [OSR.md](https://github.com/eclipse/omr/blob/master/doc/compiler/osr/OSR.md). Since the content in [OSR.md](https://github.com/eclipse/omr/blob/master/doc/compiler/osr/OSR.md) has been expanded in this PR. @gita-omr could you also help update the link in [TOC](https://github.com/eclipse-openj9/openj9/blob/master/doc/compiler/README.md) from `OSR Helpers` to `OSR Introduction`? Thank you!

```
   * On Stack Replacement (OSR)
    * [OSR Helpers (OMR)](https://github.com/eclipse/omr/blob/master/doc/compiler/osr/OSR.md)
```
==>
```
   * On Stack Replacement (OSR)
    * [OSR Introduction (OMR)](https://github.com/eclipse/omr/blob/master/doc/compiler/osr/OSR.md)
```